libraries/AP_OpticalFlow/AP_OpticalFlow_UPFLOW.cpp

```diff
@@ -0,0 +1,203 @@
+   byte0: header0 (0xFE)
+   byte1: header1 (0x0A)
+   byte2: x-motion low byte;
+   byte3: x-motion high byte;
+   byte4: y-motion low byte;
+   byte5: y-motion high byte;
+   byte6: dt low byte;
+   byte7: dt high byte;
+   byte8: reserved;
+   byte9: reserved;   
+   byte10: surface quality
+   byte11: hardware version
+   byte12:checksum
+   byte13:footer (0x55)
+ */
+
+#include <AP_HAL/AP_HAL.h>
+#include "AP_OpticalFlow_UPFLOW.h"
+#include <AP_Math/crc.h>
+#include <AP_AHRS/AP_AHRS.h>
+#include <AP_SerialManager/AP_SerialManager.h>
+#include <utility>
+#include "OpticalFlow.h"
+#include <stdio.h>
+
+#define UPFLOW_HEADER0         (uint8_t)0xFE
+#define UPFLOW_HEADER1         (uint8_t)0x0A
+#define UPFLOW_FOOTER         (uint8_t)0x55
+#define UPFLOW_PIXEL_SCALING         (1e-4)
+#define UPFLOW_TIMEOUT_SEC         0.3f
+
+extern const AP_HAL::HAL& hal;
+
+// constructor
+AP_OpticalFlow_UPFLOW::AP_OpticalFlow_UPFLOW(OpticalFlow &_frontend, AP_HAL::UARTDriver *_uart) :
+    OpticalFlow_backend(_frontend),
+    uart(_uart)
+{
+}
+
+// detect the device
+AP_OpticalFlow_UPFLOW *AP_OpticalFlow_UPFLOW::detect(OpticalFlow &_frontend)
+{
+    AP_SerialManager *serial_manager = AP::serialmanager().get_singleton();
+    if (serial_manager == nullptr) {
+        return nullptr;
+    }
+
+    // look for first serial driver with protocol defined as OpticalFlow_UPFLOW
+    AP_HAL::UARTDriver *uart = serial_manager->find_serial(AP_SerialManager::SerialProtocol_OpticalFlow_UPFLOW, 0);
+    if (uart == nullptr) {
+        return nullptr;
+    }
+
+    // we have found a serial port so use it
+    AP_OpticalFlow_UPFLOW *sensor = new AP_OpticalFlow_UPFLOW(_frontend, uart);
+    return sensor;
+}
+
+// initialise the sensor
+void AP_OpticalFlow_UPFLOW::init()
+{
+    // sanity check uart
+    if (uart == nullptr) {
+        return;
+    }
+    recv_count=0;
+    // open serial port with baud rate of 19200
+    uart->begin(19200);
+}
+
+// read latest values from sensor and fill in x,y and totals.
+void AP_OpticalFlow_UPFLOW::update(void)
+{
+    // sanity check uart
+    if (uart == nullptr) {
+        return;
+    }
+
+    // record gyro values as long as they are being used
+    // the sanity check of dt below ensures old gyro values are not used
+    if (gyro_sum_count < 1000) {
+        const Vector3f& gyro = AP::ahrs_navekf().get_gyro();
+        gyro_sum.x += gyro.x;
+        gyro_sum.y += gyro.y;
+        gyro_sum_count++;
+    }
+
+    bool phrased = false;
+    // read any available characters in the serial buffer
+    int16_t nbytes = uart->available();
+    while (nbytes-- > 0) {
+        int16_t r = uart->read();
+        if (r < 0) {
+            continue;
+        }
+        uint8_t c = (uint8_t)r;
+        if( recv_count ==0 )
+        {
+            //Header0
+            if( c != UPFLOW_HEADER0 )
+                recv_count = 0;
+            else
+            {
+                ++recv_count;
+                sum = 0;
+            }
```

Comment:
```suggestion
            if (c != UPFLOW_HEADER0) {
                recv_count = 0;
            } else {
                recv_count++;
                sum = 0;
            }
```

Similarly with the rest of this state machine.

One space after the if.

Always use braces for if-bodies

brackets on same line as if and else.

No spaces after `(`) or before `)`)

Please see our style guide for more information - and consider running `astyle` over these new files (we provide a script for applying our style requirements to a file)